### PR TITLE
[inductor][NVGEMM] Add addmm (bias-add epilogue) support with epilogue fusion

### PR DESCRIPTION
**Pull request by mlazos**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* #189841
* #189807
* #189806
* #189781
* #189780
* #189779
* #189778
* #189777
* #189776
* #189775
* __->__ #189774
* #189773
* #189805
* #189772
* #189771

Adds NVGEMM as an autotune backend for aten.addmm. The bias is baked into an
EFC (epilogue-fusion-capable) epilogue computing `bias + mat1 @ mat2`, wired in
through tuned_addmm. When the scheduler also fuses downstream pointwise ops
(gelu, residual add, scaling), the bias-add is composed into that fused
epilogue so it folds into the GEMM the way Triton's epilogue fusion does.

Review order: tuned_addmm wiring (kernel/mm.py) -> choice generation
(add_nv_universal_addmm_choices) -> the caller/buffer/kernel bias_node plumbing
(nv_universal_gemm.py, ir.py, nv_universal_gemm_kernel.py) -> subprocess
precompile + benchmark for the bias path (async_compile.py, select_algorithm.py).

Several fixes were needed to make the EFC epilogue path work end to end:
- Name the epilogue aux parameter after the bias buffer, never "C": cutlass
  LoadSrcImpl.match claims any epilogue param named "C" whose shape equals the
  output (ignoring stride), which shadows RowBroadcastImpl and silently
  mis-reads a 1D (broadcast) bias.
- Pass the epilogue to cutlass.operators as a source STRING rather than a
  callable. A callable triggers inspect.getsource() on every kernel launch,
  which raises "could not get source code" for generated modules loaded via the
  subprocess PyCodeCache.
- Fix _rewrap_efc_compiled_obj to invoke a disk-reloaded EFC artifact with
  unpacked epilogue tensors (a, b, stream, *aux) instead of pack_arguments(...);
  the reloaded kernel module has an unpacked signature. This was latent in the
  existing scheduler-fused epilogue disk-cache path (masked because loads
  usually failed and recompiled).
- Invalidate the caller's memoized output_node buffer when its name is no longer
  registered: a buffer created during a benchmark sub-context can have its
  `buf{N}` name freed when the sub-context rolls back V.graph.buffers, colliding
  with a later buffer and corrupting finalize_multi_template_buffers.

NVGEMM addmm is numerically at parity with eager/ATEN (within a bf16 ULP) and
at rough per-kernel and e2e parity with Triton on FLUX.1-dev bf16 shapes.

Test Plan:
```
# plain + fused addmm run on NVGEMM and match eager (bf16 tolerance):
python - <<'PY'
import torch
torch._inductor.config.max_autotune=True
torch._inductor.config.max_autotune_gemm_backends='NVGEMM'
M,N,K=4608,3072,3072
torch.manual_seed(0)
x=torch.randn(M,K,dtype=torch.bfloat16,device='cuda'); w=torch.randn(N,K,dtype=torch.bfloat16,device='cuda')
b=torch.randn(N,dtype=torch.bfloat16,device='cuda'); r=torch.randn(M,N,dtype=torch.bfloat16,device='cuda')
for f in [lambda:torch.addmm(b,x,w.t()),
          lambda:torch.nn.functional.gelu(torch.addmm(b,x,w.t())),
          lambda:torch.addmm(b,x,w.t())+r]:
    ref=f(); torch._dynamo.reset()
    o=torch.compile(f,fullgraph=True)()
    assert torch.allclose(o.float(),ref.float(),rtol=3e-2,atol=5e-1)
print("ok")
PY
# and a full FLUX.1-dev compile+generate with ATEN,NVGEMM.
```

Authored with assistance from Claude Code.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Comment by mlazos** [comment]
## AI-Review

> *Automated multi-agent review (2 Claude + 2 Codex + synthesis). This is AI-generated — please verify findings manually.*

---

## Critical Issues (must fix)

- **addmm input ordering mismatch between the NVUniversal choice and the autotuner** *(Codex Bug)*
  In `tuned_addmm` (`torch/_inductor/kernel/mm.py`, ~L725) the NVUniversal choice is built with caller inputs `[mat1, mat2, original_bias]`, but autotuning calls `autotune_select_algorithm(..., kernel_inputs.nodes(), ...)` with the global order `[expanded_bias, mat1, mat2]`. `benchmark_choice()` forwards that global list to the non-extern choice, so `_make_bias_run_fn()` treats `expanded_bias`/`mat1` as the GEMM operands and `mat2` as the bias.
  *Trigger:* addmm with `bias=(N,)`, `mat1=(M,K)`, `mat2=(K,N)` compiles GEMM `(M,N)@(M,K)` — fails whenever `N != M`, and silently benchmarks the wrong computation on degenerate compatible shapes.
  *Fix:* feed each choice its own `input_nodes`, or add a targeted addmm path (mirroring the existing extern addmm special case) so the NVUniversal choice benchmarks with `[mat1, mat2, original_bias]` end to end. Don't rely on the global `kernel_inputs.nodes()` order unless the choice uses that same order everywhere.

- **`_rewrap_efc_compiled_obj` calling-convention change may regress the non-bias disk-cache path** *(Claude Bug + Claude Design — consensus, 2 reviewers)*
  In `nv_universal_gemm_kernel.py`, `wrapped_launch` switched from `pack_arguments(*runtime_args)` to unpacked `*runtime_args`. This function already serves the plain (no-bias) EFC disk-cache-hit path, so the convention change is global, not bias-only. If the raw exported JIT func expects packed args, a warm on-disk cache hit will launch with the wrong arg arity — a launch failure / wrong results on a path that only triggers with a warm cache and is easy to miss in tests.
  *Action:* verify against `cutlass_api` whether the raw func signature is `(a, b, stream, packed)` or `(a, b, stream, *tensors)`. If unpacked, this is a legit latent-bug fix — make the contract explicit (assert artifact type / flag), not a comment. If they can differ, gate packing on `epilogue_args is not None`. Add a warm-cache regression test for the plain-GEMM path.

## Important Suggestions (should fix)

- **Single source of truth for the bias-add epilogue semantics** *(Claude Design #2/#3 + Codex Design #2 — consensus, 2 reviewers)*
  The same "accum + bias" operation is encoded independently in `_nvgemm_bias_add_epilogue` (callable, benchmark/precompile), `_build_bias_epilogue`/`_EPILOGUE_FN_SRC` (codegen string), and `_compose_bias_into_epilogue`, with load-bearing rules (no param named `C`, no string literals, 1D→row-broadcast) duplicated. Benchmark traces the callable while codegen emits the string; if they diverge, autotuning benchmarks a different kernel than it codegens — a correctness hazard, not just duplication. The magic `epilogue_source="nvgemm_addmm_bias_v1"` and `EpilogueArguments` construction are likewise hardcoded in two cache-key producers.
  *Action:* drive all paths from one definition (an `EpilogueSpec`-style object owning source, reads, renames, cache identity, and composition), or at minimum a shared constant + shared arg-builder, plus a test asserting the callable and string representations agree.
  *Note:* the two bug reviewers disagree on cache-split severity — Claude Bug found the disk key is keyed on `has_epilogue`/`aux_tensors`, not the source string, so a param-name drift won't split the cache. The semantic-equivalence risk (benchmark vs codegen kernel) still stands regardless.

- **Bias is double-tracked as explicit state *and* a positional "last input" side channel** *(Claude Design #1 + Codex Design #1 — consensus, 2 reviewers)*
  `bias_node` is both a field and appended to `input_nodes`, then split back off (`*gemm, bias = ...`) in ~5 sites (`_make_bias_run_fn`, `_worker_nvgemm_autotuning_precompile`, `NVUniversalGemmKernel.__init__`, `NVUniversalGemmBuffer.make_kernel_render`, caller `make_kernel_render`). This implicit positional ABI can drift if anyone adds a trailing input or reorders/serializes `input_nodes` generically. The scheduler-read-tracking reason for appending is legitimate, but the decoding shouldn't be reimplemented per site.
  *Action:* centralize the split in one helper (or a first-class addmm problem descriptor with explicit `mat1`/`mat2`/aux operands) so the convention lives in exactly one place.

## Minor Notes (consider)

- **`_compose_bias_into_epilogue` composes IR by regex-rewriting generated Python source** *(Codex Design #3)*
  Couples addmm correctness to the textual shape of `CutlassEVTCodegen` output and the `accum` param name. Prefer composing through an expression-level representation or an explicit "base accumulator" hook rather than a source rewrite. (Overlaps with the single-source-of-truth item above.)

- **`add_nv_universal_addmm_choices` accepts `alpha`/`beta` but no-ops on any value ≠ 1** *(Claude Design #5)*
  API over-promises. Either document prominently that only `alpha==beta==1` is handled (and why) or drop the params until implemented. Fine as a first cut.

- **`get_output_node` name-reuse workaround is a band-aid — file a tracking issue** *(Claude Design #6)*
  The `cached_name in V.graph.name_to_buffer` guard works around positional buffer names (`buf{len(buffers)}`) being reused after a benchmark sub-context rolls back `V.graph.buffers`. Acceptable as written, but the real fix belongs in the benchmark-context rollback/naming layer and affects all `NVUniversalGemmCaller`s. Flag it so it isn't forgotten.

---

*Investigated and cleared (no action needed):* aux-tensor ordering in `_compose_bias_into_epilogue` (epilogue inputs bound by name via `_render_epilogue_kwargs`, not position); the `bias` vs buffer-name param difference between benchmark and codegen; and the `output_node()` memoization change (`.data.data.get_name()`) — all confirmed correct by the Claude bug review.

File: torch/_inductor/codegen/nv_universal_gemm/nv_universal_gemm.py [hunk 1]
```diff
@@ -231,6 +240,92 @@ def run_kernel():
...
+                self._disk_fn_cache,
+                kernel_name,
+                disk_config_key,
+                cache_key,
+                _unwrap_efc_compiled_obj(artifact.compiled_obj),
+                dev_idx,
+            )
+
+        if self.workspace_size > 0:
+            self._workspace = torch.empty(
```
Comment (drisspg) [inline torch/_inductor/codegen/nv_universal_gemm/nv_universal_gemm.py:311]: I know we have alot of workspace stuff for auotuning not sure hwo this rolls into that

File: torch/_inductor/codegen/nv_universal_gemm/nv_universal_gemm_kernel.py [hunk 2]
```diff
@@ -698,6 +719,71 @@ def run(self, *args, stream=None, **kwargs):
...
+    CutlassEVTCodegen's convention. This deliberately avoids the name ``C``:
+    cutlass.operators' LoadSrcImpl claims any epilogue param named ``C`` whose
+    shape equals the output (ignoring stride), which shadows the row/column
+    broadcast impls and silently mis-reads a broadcast (1D) bias.
+    """
+    epilogue_fn_code = (
+        f"def _epilogue_fn(accum, {bias_name}):\n"
+        f"    D = accum + {bias_name}\n"
+        f"    return D"
+    )
```
Comment (drisspg) [inline torch/_inductor/codegen/nv_universal_gemm/nv_universal_gemm_kernel.py:747]: I wonder if we can start to merge the flex_gemm / thiese epilogue utilties for bias I think this si fine but we might we should think about this soon

**Review by drisspg — APPROVED** [review]
